plugins/radar-android-google-places/src/main/java/org.radarbase.passive.google.places/GooglePlacesManager.kt

```diff
@@ -0,0 +1,464 @@
+                    send(placesInfoTopic, GooglePlacesInfo(currentTime, currentTime, types, city, state, country, it, placeLikelihood, fromBroadcastRegistration))
+                    logger.info("Google Places data with additional info sent")
+                    return
+                }
+            }
+            send(placesInfoTopic, GooglePlacesInfo(currentTime, currentTime, types, null, null, null, placeId, placeLikelihood, fromBroadcastRegistration))
+            logger.info("Google Places data sent ")
+        }
+    }
+
+
+
+    private fun findComponent(placeInfo: FetchPlaceResponse, component: String): String? =
+         try {
+             placeInfo.place.addressComponents?.asList()?.first { address ->
+                 address.types.any { it.equals(component) }
+             }?.name
+         } catch (ex: NoSuchElementException) {
+            logger.warn("Exception while retrieving data for additional place details ", ex)
+             null
+        }
+
+    private fun handleApiException(exception: ApiException, statusCode: Int) {
+        when (statusCode) {
+            INVALID_API_CODE -> {
+                logger.error("Invalid Api key, disconnecting now")
+                disconnect()
+            }
+            LOCATION_UNAVAILABLE_CODE -> logger.warn("Location not enabled yet")
+            else -> logger.error("ApiException occurred with status code $statusCode", exception)
+        }
+    }
+
+    override fun disconnect() {
+        if (!isClosed) {
+            numOfAttempts = preferences.getInt(NUMBER_OF_ATTEMPTS_KEY, 0)
+            val currentDelay = (baseDelay + 2.0.pow(numOfAttempts)).toLong()
+            placeHandler.delay(currentDelay * 1000) {
+                if (currentDelay < maxDelay) {
+                    numOfAttempts++
+                    preferences
+                        .edit()
+                        .putInt(NUMBER_OF_ATTEMPTS_KEY, numOfAttempts).apply()
+                }
+                    super.disconnect()
+            }
+        }
+    }
+
+    fun placesFetchInterval(interval: Long, intervalUnit: TimeUnit) {
+        placesProcessor.interval(interval, intervalUnit)
+    }
+
+    fun shouldFetchPlaceId(needPlaceId: Boolean) {
+        state.shouldFetchPlaceId.set(needPlaceId)
+    }
+
+    fun shouldFetchAdditionalInfo(needAdditionalInfo: Boolean) {
+        state.shouldFetchAdditionalInfo.set(needAdditionalInfo)
+    }
+
+    fun limitByPlacesCount(limitCount: Int) {
+        state.limitByPlacesCount = limitCount
+    }
+
+    fun limitByPlacesLikelihood(limitLikelihood: Double) {
+        state.limitByPlacesLikelihood = limitLikelihood
+    }
+
+    private fun checkLocationPermissions(): Boolean =
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
+            ((ContextCompat.checkSelfPermission(service, ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED) || ContextCompat.checkSelfPermission(service, ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) }
```

Comment:
Surround the boolean assertion with `check(...) { "Location permission is missing for Google Places" }` with the permission check you have now.